implemented L1 distance & fixed L2 in binary quantization

kaancfidan opened the pull request:
This pull request is related to https://github.com/qdrant/qdrant/issues/3052.

It implements L1 distance type to be used for the matching Manhattan distance implementation in qdrant/qdrant.

I have tried to make the changes as non-disruptive as possible, but since L1 distance cannot be related to dot product as easily as L2, there had to be some branching between scoring implementations.

I have added tests and benchmarks and verified the implementation works for simple, AVX, SSE and Neon instructions.

kaancfidan committed:
implemented L1 distance

kaancfidan committed:
Added missing includes for abs

kaancfidan commented:
I just realized the changes in C implementations broke qdrant windows build, I'm trying to understand why cl.exe can't build these.

kaancfidan committed:
types simplified in SSE implementation

kaancfidan commented:
I had seen the l1_avx is not used warning, but it is being used in the benchmarks. I could remove the benchmarks and the implementation of course but do you have a better idea?

kaancfidan committed:
added allow dead_code to l1_avx

kaancfidan commented:
I have realized that this L1 distance is currently not compatible with binary quantization. I had followed mostly what had been done for L2, but I suspect L2 is also incompatible with it.

Edit: On second thought, I think it may already be compatible as the current code uses XOR in case of invert = true and the truth tables match perfectly (as it's the actual Hamming distance in the case of L1)

|   A   |   B   | Dot product | L1 |
|:-----:|:-----:|:-----------:|:--:|
| -0.5  | -0.5  |    0.25     |  0 |
| -0.5  |  0.5  |   -0.25     |  1 |
|  0.5  | -0.5  |   -0.25     |  1 |
|  0.5  |  0.5  |    0.25     |  0 |

| A | B | NXOR | XOR |
|:-:|:-:|:----:|:---:|
| 0 | 0 |   1  |  0  |
| 0 | 1 |   0  |  1  |
| 1 | 0 |   0  |  1  |
| 1 | 1 |   1  |  0  |

kaancfidan committed:
fixed neon implementation and tests

kaancfidan committed:
fixed L1 and L2 distances in binary quantization

kaancfidan commented:
I have fixed both L1 and L2 distances and added tests for both in case of binary quantization.

kaancfidan commented on quantization/src/encoded_vectors_binary.rs:
```diff
@@ -33,6 +36,8 @@ impl<TStorage: EncodedStorage> EncodedVectorsBin<TStorage> {
     ) -> Result<Self, EncodingError> {
         debug_assert!(validate_vector_parameters(orig_data.clone(), vector_parameters).is_ok());
 
+        let (alpha, _) = Self::find_alpha_offset_size_dim(orig_data.clone());
```
I have not used the offset (min value of the dataset) here, but I think it would make much sense to use it when encoding the vectors. It could widen the compatibility range of vector extractor models for binary quantization.

It would also make the implementation very symmetrical with UINT8 quantization, with the only difference between them being the quantization ratio and the actual similarity calculation.

kaancfidan replied:
I got confused before; in the case of binary vectors, offset should not be the min value but the mean of the dataset to split the data in half.

Ideally it should be an offset vector not a single f32. It's a whole other set of experiments to be run on a different branch :)

generall commented on quantization/src/encoded_vectors_binary.rs:
```diff
@@ -22,6 +24,7 @@ pub struct EncodedBinVector {
 #[derive(Serialize, Deserialize)]
 struct Metadata {
     vector_parameters: VectorParameters,
+    alpha: f32,
```
this probably should be optional, so previous versions could be loaded as well

kaancfidan replied:
I think it would be better to add backwards compatibility test cases like this, but I'm not familiar enough with the code base and don't know how the data is persisted.

kaancfidan replied:
I have changed it to `Option<f32>` without adding a test case. Should there be added storage compatibility test cases on qdrant side?

timvisee replied:
We might want `#[serde(default)]` before it as well.

kaancfidan committed:
binary quantization alpha is now optional

kaancfidan commented:
> I have fixed both L1 and L2 distances and added tests for both in case of binary quantization.

I have run a benchmark with real data using Euclid + binary quantization (using qdrant itself). It did not work well. I need to investigate and understand the difference between the toy test data and the real one.

I'll also check L1 tomorrow. Meanwhile, I have set the status to draft again.

timvisee commented on quantization/tests/metrics.rs:
```diff
@@ -9,3 +9,11 @@ pub fn l2_similarity(v1: &[f32], v2: &[f32]) -> f32 {
 }
+
+pub fn l1_similarity(v1: &[f32], v2: &[f32]) -> f32 {
+    v1.iter()
+        .copied()
+        .zip(v2.iter().copied())
+        .map(|(a, b)| (a - b).abs())
+        .sum()
```
We don't have to copy explicitly:

```suggestion
    v1.iter().zip(v2).map(|(a, b)| (a - b).abs()).sum()
```

We can do the same with the `l2_similarity` function.

timvisee commented on quantization/tests/test_avx2.rs:
```diff
@@ -84,4 +84,41 @@ mod tests {
+        let mut rng = rand::rngs::StdRng::seed_from_u64(42);
+        let mut vector_data: Vec<Vec<f32>> = Vec::new();
+        for _ in 0..vectors_count {
+            let vector: Vec<f32> = (0..vector_dim).map(|_| rng.gen_range(-1.0..1.0)).collect();
```
```suggestion
            let vector: Vec<f32> = (0..vector_dim).map(|_| rng.gen_range(-1.0..=1.0)).collect();
```

Nitpick: we might want `..=` here to be inclusive.

If so, there are quite a few other places where we want this as well.

kaancfidan replied:
I had actually added this range because one of my initial attempts of implementation got false positive tests when all the vectors were in [0, 1].

I personally don't think it matters if it includes 1.0 or not in this context, but I could change it of course.

timvisee commented on quantization/src/encoded_vectors_binary.rs:
```diff
@@ -123,29 +129,53 @@ impl<TStorage: EncodedStorage> EncodedVectorsBin<TStorage> {
-            zeros_count as f32 - xor_product as f32
+            match self.metadata.vector_parameters.distance_type {
+                DistanceType::Dot => zeros_count as f32 - xor_product as f32,
+                DistanceType::L1 => 1.0 / (alpha * xor_product as f32),
+                DistanceType::L2 => 1.0 / (alpha * alpha * xor_product as f32),
```
What if the `/ rhs` ever becomes `0.0`?

kaancfidan replied:
Yes, I also caught that, I have changed this locally actually to `dim*exp(-x)` so it ranges between dim -> 0.

I also dropped the whole alpha estimation as Dot already assumes [-1, 1] range and I don't want to change it. Decided to hardcode alpha=2 for L1 and L2.

timvisee commented on quantization/src/encoded_vectors.rs:
```diff
@@ -37,6 +38,7 @@ impl DistanceType {
     pub fn distance(&self, a: &[f32], b: &[f32]) -> f32 {
         match self {
             DistanceType::Dot => a.iter().zip(b.iter()).map(|(a, b)| a * b).sum(),
+            DistanceType::L1 => a.iter().zip(b.iter()).map(|(a, b)| (a - b).abs()).sum(),
             DistanceType::L2 => a.iter().zip(b.iter()).map(|(a, b)| (a - b) * (a - b)).sum(),
```
Though fine, `.iter()` is redundant here:

```suggestion
            DistanceType::Dot => a.iter().zip(b).map(|(a, b)| a * b).sum(),
            DistanceType::L1 => a.iter().zip(b).map(|(a, b)| (a - b).abs()).sum(),
            DistanceType::L2 => a.iter().zip(b).map(|(a, b)| (a - b) * (a - b)).sum(),
```

timvisee commented on quantization/src/encoded_vectors_u8.rs:
```diff
@@ -335,43 +394,94 @@ impl<TStorage: EncodedStorage> EncodedVectors<EncodedQueryU8> for EncodedVectors
+    unsafe {
+        let mut score = 0i32;
+        for i in 0..actual_dim {
+            score += ((*q_ptr.add(i) as i32) - (*v_ptr.add(i) as i32)).abs();
```
We might prefer [`abs_diff`](https://doc.rust-lang.org/std/primitive.i32.html#method.abs_diff) here instead.

```suggestion
            score += (*q_ptr.add(i) as i32).abs_diff(*v_ptr.add(i) as i32);
```

timvisee reviewed (commented):
Did a first review pass with some minor comments. Thank you very much for your time and effort to implement this! We'll definitely take a proper look at this.

kaancfidan committed:
scoring direction fixes and some refactoring based on review

kaancfidan commented:
> > I have fixed both L1 and L2 distances and added tests for both in case of binary quantization.
> 
> I have run a benchmark with real data using Euclid + binary quantization (using qdrant itself). It did not work well. I need to investigate and understand the difference between the toy test data and the real one.
> 
> I'll also check L1 tomorrow. Meanwhile, I have set the status to draft again.

I have been banging my head on this all day. 

I am pretty sure L1 and L2 distances and inverses (similarities) are calculated properly. I have gone through unit tests, added integration tests that use binary and scalar quantizations on qdrant...etc. they all work as I expect.

Nevertheless, when I run our benchmark image vector dataset with this implementation, I still get 0% accuracy with L1 and L2 using binary quantization while dot product gets about 84%. 

I will let it go if I can justify it as "our problem is just not compatible with binary quantization" but the fact that dot product works annoys me to no end. I am open to suggestions.

kaancfidan committed:
simplified calculate_metric

kaancfidan commented:
OK, I had an epiphany where I understood that the binary `calculate_metric` method does not need to actually calculate the correct metric (dot, L1 or L2) but just has to approximate their behavior and put the vectors in the same order. The actual score is being calculated when rescoring anyway. 

I have relaxed the L1 and L2 binary tests to just check the similarity order but not care about the actual score values. This way I could see that the previous implementation with dot product is still on point for both L1 and L2, but it was just inverse of what it was supposed to be. The many layers of inversing confused me too.

I apologize to the reviewers that they had to go through my learning experience with me. :)

I will run the large benchmark with this latest version as soon as I can build a docker image.

edit: aaaand we have a winner! 🍾

kaancfidan committed:
comment fixes and reordering for clarity

timvisee committed:
Merge imports

timvisee committed:
Align truth table

timvisee committed:
Remove redundant copies

timvisee reviewed (approved):
Nice work!

I must admit that I didn't look into the C implementations. Maybe @IvanPleshkov can take a look.

IvanPleshkov commented on quantization/cpp/sse.c:
```diff
@@ -53,3 +62,50 @@ EXPORT uint64_t impl_xor_popcnt_sse(
+    // Horizontal sum using the macro
+    HSUM128_EPI16(sum_epi32, sum);
+
+    // Sum the remaining elements
```
Not required. dimension is guaranteed to div 16

IvanPleshkov commented on quantization/cpp/neon.c:
```diff
@@ -63,3 +64,47 @@ EXPORT uint64_t impl_xor_popcnt_neon(
+    uint32_t sum = vget_lane_u32(sum64_high, 0);
+
+    // Sum the remaining elements
+    for (uint32_t i = m; i < dim; ++i) {
```
Not required. dimension is guaranteed to div 16

IvanPleshkov commented on quantization/cpp/avx2.c:
```diff
@@ -49,3 +59,47 @@ EXPORT float impl_score_dot_avx(
+    HSUM256_EPI32(sum_epi32, sum);
+
+    // Sum the remaining elements
+    for (uint32_t i = m; i < dim; ++i) {
```
Looks incorrect. Before `abs` substraction seems to be processed with `uint8_t` type.
Moreover, dimension is guaranteed to div 16. You may use SSE instead for the last 16 elements.

kaancfidan replied:
Ah, I had looked at the dot implementation and had not understood the use of shorts.

I'll match the implementation.

kaancfidan replied:
May I also add a comment to the dot implementation for future readers to explicitly understand that it is guaranteed to be a multiple of 16?

IvanPleshkov replied:
Yep, you may add it. You may cover by any comments as new reader of this repo

IvanPleshkov commented on quantization/tests/test_pq.rs:
```diff
@@ -81,6 +81,38 @@ mod tests {
     }
 
+    #[test]
+    fn test_pq_l1() {
```
I like that product quantization tests presented here because PQ affect is non explicit. Just adding L1 into `DistanceType::distance` should be enough.

kaancfidan replied:
I added the test first and tried to understand for a while why it was not failing 😅

IvanPleshkov commented on quantization/src/encoded_vectors_binary.rs:
```diff
@@ -120,30 +121,38 @@ impl<TStorage: EncodedStorage> EncodedVectorsBin<TStorage> {
+        let dim = self.metadata.vector_parameters.dim as f32;
+        let zeros_count = dim - xor_product;
+
+        match (
```
thanks for fixing `invert` parameter! It seems you may do this part of code simpler because with `invert=true` you can just multiply result to `-1`

kaancfidan replied:
I like my overly explicit way more actually, I want to keep it if that's OK for you? 

It shows that Dot product is a similarity metric but L1 and L2 are distances that need to be inverted to become similarity metrics.

IvanPleshkov replied:
OK for me

IvanPleshkov commented on quantization/src/encoded_vectors_binary.rs:
```diff
@@ -120,30 +121,38 @@ impl<TStorage: EncodedStorage> EncodedVectorsBin<TStorage> {
-
         // Dot product in a range [-1; 1] is approximated by NXOR in a range [0; 1]
+        // L1 distance in range [-1; 1] (alpha=2) is approximated by alpha*XOR in a range [0; 1]
+        // L2 distance in range [-1; 1] (alpha=2) is approximated by alpha*sqrt(XOR) in a range [0; 1]
```
Why `sqrt(XOR)` here?

IvanPleshkov replied:
I got it, `sqrt(1)=1` and `sqrt(0)=0`

kaancfidan replied:
well sqrt is there in the L2 definition, if you are not taking the sqrt at a later step you are actually calculating L2^2.

I had actually confirmed these approximations through tests, but decided to forgoe te actual value approximations for calculation simplicity. alpha or sqrt does not change the order.

IvanPleshkov commented:
It's a great work! I really wondered! Impressed with correct scalar quantization `alpha` and `multiplier` values and fixed for non-L1 related bugs. I just added some requests for C code. Everything else looks perfect!

kaancfidan committed:
AVX, SSE and Neon fixes for uint8
